docs/contributing/Style-guide-for-Flutter-repo.md

```diff
@@ -831,21 +831,60 @@ really is, and tends to encourage "spaghetti" code (where distant components ref
 than "lasagna" code (where each section of the code is cleanly layered and separable).
 
 
-### Avoid using `extension`.
+### Avoid using `extension` methods.
 
 Extension methods are confusing to document and discover. To an end developer,
 they appear no different than the built-in API of the class, and discovering
 the documentation and implementation of an extension is more challenging than
-for class members.
+for class members. Prefer instead adding methods directly to relevant classes.
 
-Prefer instead adding methods directly to relevant classes. If that is not
-possible, create a method that clearly identifies what object(s) it works with
-and is part of.
+Only use `extension` methods when discoverability is not a concern:
 
-(A rare exception can be made for extensions that provide temporary workarounds
+- Extension methods might be implemented as temporary workarounds
 when deprecating features. In those cases, however, the extensions and all their
 members must be deprecated in the PR that adds them, and they must be removed
 in accordance with our deprecation policy.)
+- Additionally, private class fields
+[cannot be safely accessed](https://github.com/dart-lang/sdk/issues/51641)
+if the receiver is a subtype that implements the interface.\
+In these cases, it's better to house these fields in an `extension`
+or to restructure them into private top-level functions if possible.
+
+
+### Use extension types to improve an API surface
+
+"Extension types" in Dart are distinct from "extension methods":
+
+```dart
+extension A on B {
+  void c() {}
+}
+```
+
+ * `A` is an _extension_.
+ * `B` is the _type_ (to be specific, it's the "`on` type").
+ * `c()` is an _extension method_.
+
+```dart
+extension type A(B b) implements C {
+  void d() {}
+}
+```
+
+ * `A` is an _extension type_.
+ * `B` is the _representation type_ (and `b` is the "representation field").
+ * `C` is the _supertype_ (optionally added using an "`implements` clause").
+ * `d()` is a member of `A`'s interface.\
+   (to prevent confusion, avoid referring to it as an "extension method").
+
+Rather than adding fields to an existing class, an `extension type` creates
+a separate interface. This can prevent [API oceans](#avoid-exposing-api-oceans)
+without any risk of name overlap, and while avoiding the performance cost of
+a wrapper class.
+
+Prefer specifying a non-nullable supertype (e.g. `implements Object`) when
+the representation type is non-nullable: this improves static analysis for
+null-checks.
```

Comment:
This reads more like a description of what extension types are, than advice for when it's appropriate to use them. (The heading, "Use extension types to improve an API surface", taken literally, suggests using them all the time, which is probably not what is intended.)

I would probably replace this whole section with a parenthetical sentence in the previous section that just says something like "(Unlike extension methods, the use of extension _types_ is fine.)", maybe with a link to the Dart documentation on what extension types are, what they're useful for, etc.

Reply:
Thanks again for bringing up that the heading should be changed.

As far as including a description, it seems that the dart.dev pages for [**extension methods**](https://dart.dev/language/extension-methods) and [**extension types**](https://dart.dev/language/extension-types) are geared toward describing rather than differentiating.
I believe including a distinction in the style guide would likely help contributors better understand how they should apply to the Flutter framework, though perhaps there would be a better place for this.

Reply:
The style guide is meant to teach people what we've learned are best practices, it's not meant to teach people Dart. We should assume people know what extension methods and extension types are, and if we feel we need to say something about extension types at all, it should be about when they're appropriate with concrete examples. But we don't _need_ to mention every Dart feature, only those where the choice of using it or not using it, or how to use it, is not necessarily obvious (especially as shown by people submitting PRs that misuse them).

Reply:
Apologies for the late response here—you bring up some good points but overall I'm not completely on board.

I 100% agree that the information we choose to put in the style guide should reflect what problems or confusion we notice in PRs. I also agree with the style guide's advice regarding extension methods:

> Prefer instead adding methods directly to relevant classes.

Since the dart.dev docs don't discourage either `extension` or `extension type`, there isn't much of a reason to make an explicit effort to differentiate between them, but based on interactions I've had within PR discussions, I do feel that it'd be helpful to draw this distinction here.
I'm interested to hear other perspectives on this, whether it's through this comment thread or the tentative upcoming Dash Forum :)